### PR TITLE
[CCUBE-1614][JON]Add withinDays condition to FEE

### PR DESCRIPTION
**Pull request by hhung2710**
**Changes**
Validate if value is within x days more or less from now or specific date
Add validation rule to date-field also as a general conditional rendering rule 

-   [keep | delete] branch

**Changelog entry**

-   [BREAKING | WARNING | remove if none] Description of change

**Additional information**

-   You may refer to this [CCUBE-1614](https://jira.ship.gov.sg/browse/CCUBE-1614)
-   Any other information you would like to include

### CODE REVIEW

File: src/components/fields/date-field/types.ts [hunk 1]
```diff
@@ -11,11 +11,15 @@ export interface IDateFieldValidationRule extends IYupValidationRule {
 	minDate?: string | undefined;
 	maxDate?: string | undefined;
 	excludedDates?: string[] | undefined;
+	withinDays?: {
+		numberOfDays: number;
+		specificDate?: string | undefined;
```
Comment (weili-govtech) [inline src/components/fields/date-field/types.ts:16]:
```suggestion
		fromDate?: string | undefined;
```

File: src/context-providers/yup/custom-conditions/index.ts [hunk 2]
```diff
@@ -41,3 +43,38 @@ YupHelper.addCondition("mixed", "equalsField", (values: unknown[], matches: unkn
...
+			const today = LocalDate.now();
+			if (numberOfDays >= 0) {
+				startDate = today;
+				endDate = today.plusDays(numberOfDays);
+			} else {
+				startDate = today.plusDays(numberOfDays);
+				endDate = today;
+			}
+		}
+		return !localDate.isBefore(startDate) && !localDate.isAfter(endDate);
```
Comment (weili-govtech) [inline src/context-providers/yup/custom-conditions/index.ts:78]: can we share the logic with `date-field`?

File: src/context-providers/yup/types.ts [hunk 3]
```diff
@@ -78,6 +78,11 @@ export interface IYupConditionalValidationRule extends IYupRule {
 export interface IYupRenderRule extends IYupRule {
 	filled?: boolean | undefined;
 	shown?: boolean | undefined;
+	withinDays?: {
+		numberOfDays: number;
+		specificDate?: string | undefined;
+		dateFormat?: string;
```
Comment (weili-govtech) [inline src/context-providers/yup/types.ts:84]:
```suggestion
		dateFormat?: string | undefined;
```

File: src/context-providers/yup/custom-conditions/index.ts [hunk 4]
```diff
@@ -41,3 +43,38 @@ YupHelper.addCondition("mixed", "equalsField", (values: unknown[], matches: unkn
...
+	(
+		value: string,
+		withinDays: {
+			numberOfDays: number;
+			specificDate?: string;
+			dateFormat?: string;
+		}
+	) => {
+		if (!value) return true;
+		const { numberOfDays, specificDate, dateFormat = "yyyy-MM-dd" } = withinDays;
```
Comment (weili-govtech) [inline src/context-providers/yup/custom-conditions/index.ts:59]:
```suggestion
		const { numberOfDays, specificDate, dateFormat = "uuuu-MM-dd" } = withinDays;
```

use `uuuu` instead of `yyyy`: https://stackoverflow.com/questions/41177442/uuuu-versus-yyyy-in-datetimeformatter-formatting-pattern-codes-in-java

File: src/stories/3-fields/date-field/date-field.stories.tsx [hunk 5]
```diff
@@ -77,6 +77,18 @@ const meta: Meta = {
 				type: "boolean",
 			},
 		},
+		withinDays: {
+			description: "Validate if value is within x days from now or specific date",
```
Comment (weili-govtech) [inline src/stories/3-fields/date-field/date-field.stories.tsx:81]: explain that if a specific date is provided, the format will follow the `dateFormat` option

File: src/context-providers/yup/custom-conditions/index.ts [hunk 6]
```diff
@@ -41,3 +43,38 @@ YupHelper.addCondition("mixed", "equalsField", (values: unknown[], matches: unkn
 			return isEqual(values, fn.parent[`${matches}`]);
 	}
 });
+
+YupHelper.addCondition(
```
Comment (weili-govtech) [inline src/context-providers/yup/custom-conditions/index.ts:47]: can we have a new set of conditional rendering stories for dates?

**Review by weili-govtech — CHANGES_REQUESTED** [review]
can we have unit tests for the date-field and conditional rendering?

File: src/components/shared/error-messages.tsx [hunk 7]
```diff
@@ -21,6 +21,10 @@ export const ERROR_MESSAGES = {
 		MAX_DATE: (date: string) => `Date cannot be later than ${date}.`,
 		DISABLED_DATES: "Date is not allowed.",
 		INVALID: "Invalid date.",
+		WITHIN_DAYS: (numberOfDays: number, fromDate?: string | undefined) =>
+			`Within ${numberOfDays} days ${numberOfDays ? `more` : `less`} ${
+				fromDate ? `from ${fromDate}` : `from today`
+			}`,
```
Comment (weili-govtech) [inline src/components/shared/error-messages.tsx:27]:
```suggestion
		WITHIN_DAYS: (numberOfDays: number, fromDate?: string | undefined) =>
			`Within ${numberOfDays} days ${numberOfDays ? `more` : `less`} ${
				fromDate ? `from ${fromDate}` : `from today`
			}`,
```
```suggestion
		WITHIN_DAYS: (numberOfDays: number, fromDate?: string | undefined) => {
			const unsignedDays = Math.abs(numberOfDays);
			return `Date must be within ${unsignedDays} ${unsignedDays !== 1 ? "days" : "day"} ${
				numberOfDays > 0 ? "from" : "prior to"
			} ${fromDate ? ` ${fromDate}` : " today"}.`;
		},
```

File: src/components/fields/date-field/date-field.tsx [hunk 8]
```diff
@@ -111,20 +112,33 @@ export const DateField = (props: IGenericFieldProps<IDateFieldSchema>) => {
 						if (!isValidDate(value) || !excludedDatesRule) return true;
 						return !excludedDatesRule["excludedDates"].includes(value);
 					}
+				)
+				.test(
+					"within-days",
+					withinDaysRule?.errorMessage || ERROR_MESSAGES.DATE.WITHIN_DAYS(withinDaysRule?.["withinDays"]),
+					(value) => {
+						if (!isValidDate(value) || !withinDaysRule) return true;
```
Comment (qroll) [inline src/components/fields/date-field/date-field.tsx:120]: does this rule support `0` days? it would be falsy and pass the check

File: src/components/fields/date-field/types.ts [hunk 9]
```diff
@@ -11,11 +11,15 @@ export interface IDateFieldValidationRule extends IYupValidationRule {
 	minDate?: string | undefined;
 	maxDate?: string | undefined;
 	excludedDates?: string[] | undefined;
+	withinDays?: {
+		numberOfDays: number;
+		fromDate?: string | undefined;
+	};
```
Comment (weili-govtech) [inline src/components/fields/date-field/types.ts:17]:
```suggestion
	withinDays?: {
		numberOfDays: number;
		fromDate?: string | undefined;
	} | undefined;
```

File: src/context-providers/yup/types.ts [hunk 10]
```diff
@@ -78,6 +78,11 @@ export interface IYupConditionalValidationRule extends IYupRule {
 export interface IYupRenderRule extends IYupRule {
 	filled?: boolean | undefined;
 	shown?: boolean | undefined;
+	withinDays?: {
+		numberOfDays: number;
+		fromDate?: string | undefined;
+		dateFormat?: string;
+	};
```
Comment (weili-govtech) [inline src/context-providers/yup/types.ts:85]: create a new interface and apply it to `IYupRenderRule` and `IDateFieldValidationRule` and `calculateWithinDaysRange()`

```
export interface IWithinDaysRule {
	numberOfDays: number;
	fromDate?: string | undefined;
	dateFormat?: string | undefined;
}
```

File: src/utils/date-time-helper.ts [hunk 11]
```diff
@@ -57,4 +57,43 @@ export namespace DateTimeHelper {
 			return undefined;
 		}
 	}
+
+	export function calculateWithinDaysRange(withinDays: {
```
Comment (weili-govtech) [inline src/utils/date-time-helper.ts:61]: can we have unit test for this function?

File: src/__tests__/components/fields/date-field/date-field.spec.tsx [hunk 12]
```diff
@@ -182,14 +182,15 @@ describe(UI_TYPE, () => {
...
-		${"excluded-dates"} | ${{ excludedDates: ["2022-01-02"] }} | ${["02", "01", "2022"]} | ${["03", "01", "2022"]}
+		condition           | config                                 | invalid                 | valid
+		${"future"}         | ${{ future: true }}                    | ${["01", "01", "2022"]} | ${["02", "01", "2022"]}
+		${"past"}           | ${{ past: true }}                      | ${["01", "01", "2022"]} | ${["12", "31", "2021"]}
+		${"non-future"}     | ${{ notFuture: true }}                 | ${["02", "01", "2022"]} | ${["01", "01", "2022"]}
+		${"non-past"}       | ${{ notPast: true }}                   | ${["31", "12", "2021"]} | ${["01", "01", "2022"]}
+		${"min-date"}       | ${{ minDate: "2022-01-02" }}           | ${["01", "01", "2022"]} | ${["02", "01", "2022"]}
+		${"max-date"}       | ${{ maxDate: "2022-01-02" }}           | ${["03", "01", "2022"]} | ${["02", "01", "2022"]}
+		${"excluded-dates"} | ${{ excludedDates: ["2022-01-02"] }}   | ${["02", "01", "2022"]} | ${["03", "01", "2022"]}
+		${"within-days"}    | ${{ withinDays: { numberOfDays: 7 } }} | ${["09", "01", "2022"]} | ${["02", "01", "2022"]}
```
Comment (weili-govtech) [inline src/__tests__/components/fields/date-field/date-field.spec.tsx:193]: add test for negative value too

File: src/context-providers/yup/custom-conditions/index.ts (hunk 6 above)
Comment (weili-govtech) [reply src/context-providers/yup/custom-conditions/index.ts:46]: I think you missed this comment? I will like to have unit tests and stories for this too?

you can have a new set of conditional rendering stories for dates, can refer to `conditional-rendering-array.stories.tsx`.

add unit tests for this conditional rendering rule to `conditional-renderer.spec.tsx`

File: src/stories/3-fields/date-field/date-field.stories.tsx (hunk 5 above)
Comment (weili-govtech) [reply src/stories/3-fields/date-field/date-field.stories.tsx]: sorry, my bad, pls remove the `withinDays` from the stories because it is a validation rule but these descriptions are for props.

**Review by weili-govtech — CHANGES_REQUESTED** [review]

File: src/stories/3-fields/date-field/date-field.stories.tsx [hunk 13]
```diff
@@ -284,3 +284,36 @@ HideInputKeyboard.args = {
...
+	validation: [
+		{ required: true },
+		{
+			withinDays: {
+				numberOfDays: 10,
+				fromDate: "2024-11-09",
+			},
+		},
+	],
+};
```
Comment (weili-govtech) [inline src/stories/3-fields/date-field/date-field.stories.tsx:319]: add another story to demonstrate `dateFormat` option

File: src/context-providers/yup/types.ts [hunk 14]
```diff
@@ -48,6 +49,7 @@ interface IYupRule {
 	excludes?: unknown | undefined;
 	uinfin?: boolean | undefined;
 	equalsField?: unknown | undefined;
+	withinDays?: IWithinDaysRule | undefined;
```
Comment (weili-govtech) [inline src/context-providers/yup/types.ts:52]: `withinDays` should not be added to `IYupRule` because it will make this validation rule available to all fields but we just want it to be applied to `date-field`. (current changes to `date-field` work already)

we just want this rule to be available for conditional rendering, so it should be moved to `IYupConditionalValidationRule`.

File: src/components/fields/date-field/types.ts [hunk 15]
```diff
@@ -11,11 +12,12 @@ export interface IDateFieldValidationRule extends IYupValidationRule {
 	minDate?: string | undefined;
 	maxDate?: string | undefined;
 	excludedDates?: string[] | undefined;
+	withinDays?: IWithinDaysRule;
```
Comment (weili-govtech) [inline src/components/fields/date-field/types.ts:15]: for all optional properties in FEE and DS, we apply `undefined` union. this is to cater to consumers who enabled `exactOptionalPropertyTypes` which will cause optional properties without the undefined union to be erroneous.

```suggestion
	withinDays?: IWithinDaysRule | undefined;
```

File: src/components/fields/date-field/types.ts [hunk 16]
```diff
@@ -11,11 +12,12 @@ export interface IDateFieldValidationRule extends IYupValidationRule {
...
 }
 
 export interface IDateFieldSchema<V = undefined>
 	extends IBaseFieldSchema<"date-field", V, IDateFieldValidationRule>,
-		TComponentOmitProps<DateInputProps, "minDate" | "maxDate" | "disabledDates"> {
+		TComponentOmitProps<DateInputProps, "minDate" | "maxDate" | "disabledDates" | "withinDays"> {
```
Comment (weili-govtech) [inline src/components/fields/date-field/types.ts:20]: not needed, we use `TComponentOmitProps` to exclude props from DS interfaces but `withinDays` do not exist there in the first place

File: src/components/fields/date-field/types.ts (hunk 15 above)
Comment (weili-govtech) [reply src/components/fields/date-field/types.ts:15]: I also notice the intention to use `IDateFieldSchema`'s `dateFormat` instead of the one from `withinDays`, I think this is fine but we will need to omit `dateFormat` prop from `IWithinDaysRule` here for clarity.

File: src/components/fields/date-field/date-field.tsx [hunk 17]
```diff
@@ -111,20 +112,35 @@ export const DateField = (props: IGenericFieldProps<IDateFieldSchema>) => {
...
 
 		// set minDate / maxDate / disabledDates props
+		const withinDaysRange =
+			withinDaysRule?.["withinDays"] &&
+			DateTimeHelper.calculateWithinDaysRange({ ...withinDaysRule["withinDays"], dateFormat });
 		const minDateProp = getLatestDate([
 			minDate,
 			futureRule?.["future"] && LocalDate.now().plusDays(1),
 			notPastRule?.["notPast"] && LocalDate.now(),
+			withinDaysRule?.["withinDays"] && withinDaysRange.startDate,
```
Comment (weili-govtech) [inline src/components/fields/date-field/date-field.tsx:137]: `withinDaysRange` already got the truthy check for `withinDaysRule?.["withinDays"]`

```suggestion
			withinDaysRange?.startDate,
```

File: src/components/fields/date-field/date-field.tsx [hunk 18]
```diff
@@ -111,20 +112,35 @@ export const DateField = (props: IGenericFieldProps<IDateFieldSchema>) => {
...
 		]);
 		const maxDateProp = getEarliestDate([
 			maxDate,
 			pastRule?.["past"] && LocalDate.now().minusDays(1),
 			notFutureRule?.["notFuture"] && LocalDate.now(),
+			withinDaysRule?.["withinDays"] && withinDaysRange.endDate,
```
Comment (weili-govtech) [inline src/components/fields/date-field/date-field.tsx:143]:
```suggestion
			withinDaysRange?.endDate,
```

File: src/utils/date-time-helper.ts [hunk 19]
```diff
@@ -57,4 +58,35 @@ export namespace DateTimeHelper {
...
+
+	export function checkWithinDays(value: string, withinDays: IWithinDaysRule) {
+		if (!value) return true;
+		const { dateFormat = "uuuu-MM-dd" } = withinDays;
+		const localDate = DateTimeHelper.toLocalDateOrTime(value, dateFormat, "date");
+		if (!localDate) return false;
+
+		const { startDate, endDate } = calculateWithinDaysRange(withinDays);
+		return !localDate.isBefore(startDate) && !localDate.isAfter(endDate);
+	}
```
Comment (weili-govtech) [inline src/utils/date-time-helper.ts:91]: can we have unit tests for these 2 functions?

File: src/stories/2-frontend-engine/conditional-rendering/conditional-rendering.stories.tsx [hunk 20]
```diff
@@ -211,3 +211,33 @@ OrConditions.args = {
...
+			},
+		],
+	},
+	field2: {
+		label: "Field 2",
+		uiType: "text-field",
+		showIf: [{ field1: [{ filled: true }] }],
+		validation: [{ required: true }],
+	},
+};
```
Comment (weili-govtech) [inline src/stories/2-frontend-engine/conditional-rendering/conditional-rendering.stories.tsx:243]:
```suggestion
export const WithinDaysConditions = Template.bind({});
WithinDaysConditions.args = {
	intro: {
		uiType: "div",
		className: "margin-bottom-1",
		children: "Show field 2 as long as field 1 is within 5 days from today",
	},
	field1: {
		label: {
			mainLabel: "Field 1",
		},
		uiType: "date-field",
	},
	field2: {
		label: "Field 2",
		uiType: "text-field",
		showIf: [{ field1: [{ filled: true }, { withinDays: { numberOfDays: 5 } }] }],
		validation: [{ required: true }],
	},
};

```

File: src/__tests__/components/fields/date-field/date-field.spec.tsx [hunk 21]
```diff
@@ -182,14 +182,17 @@ describe(UI_TYPE, () => {
...
+		${"non-future"}     | ${{ notFuture: true }}                                         | ${["02", "01", "2022"]} | ${["01", "01", "2022"]}
+		${"non-past"}       | ${{ notPast: true }}                                           | ${["31", "12", "2021"]} | ${["01", "01", "2022"]}
+		${"min-date"}       | ${{ minDate: "2022-01-02" }}                                   | ${["01", "01", "2022"]} | ${["02", "01", "2022"]}
+		${"max-date"}       | ${{ maxDate: "2022-01-02" }}                                   | ${["03", "01", "2022"]} | ${["02", "01", "2022"]}
+		${"excluded-dates"} | ${{ excludedDates: ["2022-01-02"] }}                           | ${["02", "01", "2022"]} | ${["03", "01", "2022"]}
+		${"within-days"}    | ${{ withinDays: { numberOfDays: 7 } }}                         | ${["09", "01", "2022"]} | ${["02", "01", "2022"]}
+		${"within-days"}    | ${{ withinDays: { numberOfDays: -7 } }}                        | ${["02", "01", "2022"]} | ${["31", "12", "2021"]}
+		${"within-days"}    | ${{ withinDays: { numberOfDays: 5, fromDate: "2022-01-05" } }} | ${["01", "01", "2022"]} | ${["06", "01", "2022"]}
```
Comment (weili-govtech) [inline src/__tests__/components/fields/date-field/date-field.spec.tsx:195]: - differentiate between the tests
- add test for date format

```suggestion
		${"within-days (future)"}        | ${{ withinDays: { numberOfDays: 7 } }}                         | ${["09", "01", "2022"]} | ${["02", "01", "2022"]}
		${"within-days (past)"}          | ${{ withinDays: { numberOfDays: -7 } }}                        | ${["02", "01", "2022"]} | ${["31", "12", "2021"]}
		${"within-days (from specific date)"} | ${{ withinDays: { numberOfDays: 5, fromDate: "2022-01-05" } }} | ${["01", "01", "2022"]} | ${["06", "01", "2022"]}
```

File: src/__tests__/components/elements/wrapper/conditional-renderer.spec.tsx [hunk 22]
```diff
@@ -186,6 +195,47 @@ describe("conditional-renderer", () => {
...
+			await changeDate(invalid[0], invalid[1], invalid[2]);
+			fireEvent.click(getField("button", "Done"));
+			expect(getFieldTwo(true)).not.toBeInTheDocument();
+
+			await changeDate(valid[0], valid[1], valid[2]);
+			fireEvent.click(getField("button", "Done"));
+			expect(getFieldTwo()).toBeInTheDocument();
+		});
+	});
+
```
Comment (weili-govtech) [inline src/__tests__/components/elements/wrapper/conditional-renderer.spec.tsx:238]: - wrap these tests in a new `within-days condition` test suites after `shown condition` test suite
- change existing `describe.each` to `it.each`
- update test label
- add test for date format

so it should look something like this

```
describe("shown condition", () => {
	// ...
});

// do this
describe("within-days condition", () => {
  	describe.each`
		condition                             | config                                                         | invalid                 | valid
		${"within-days (future)"}             | ${{ withinDays: { numberOfDays: 7 } }}                         | ${["09", "01", "2022"]} | ${["02", "01", "2022"]}
		${"within-days (past)"}               | ${{ withinDays: { numberOfDays: -7 } }}                        | ${["02", "01", "2022"]} | ${["31", "12", "2021"]}
		${"within-days (from specific date)"} | ${{ withinDays: { numberOfDays: 5, fromDate: "2022-01-05" } }} | ${["01", "01", "2022"]} | ${["06", "01", "2022"]}
		// add test for date format
	`("$condition validation", ({ condition, config, invalid, valid }) => {
		// ...
	});
});
```

**Review by weili-govtech — CHANGES_REQUESTED** [review]

File: src/components/fields/date-field/date-field.tsx [hunk 23]
```diff
@@ -48,36 +48,44 @@ export const DateField = (props: IGenericFieldProps<IDateFieldSchema>) => {
 		const minDateRule = validation?.find((rule) => "minDate" in rule);
 		const maxDateRule = validation?.find((rule) => "maxDate" in rule);
 		const excludedDatesRule = validation?.find((rule) => "excludedDates" in rule);
+		const withinDaysRule = validation?.find((rule) => "withinDays" in rule);
+		// determine the current date format to use
+		let currentDateFormat = dateFormat;
+		if (withinDaysRule?.["withinDays"].dateFormat) {
+			currentDateFormat = withinDaysRule?.["withinDays"].dateFormat;
+			setDateFormat(currentDateFormat);
+		}
```
Comment (weili-govtech) [inline src/components/fields/date-field/date-field.tsx:57]: apologies, after some considerations, I think it is better to exclude `dateFormat` from the `withinDays` validation rule in date-field. my concern
- having the `dateFormat` in `IDateFieldValidationRule` override the one `IDateFieldSchema` is confusing, developers will not be aware of it
- the alternative of making developers ensure `dateFormat` in both `IDateFieldSchema` and `IDateFieldValidationRule` causes confusion because it is a hidden logic

so I propose to
- exclude `dateFormat` from the `withinDays` validation rule in date-field
- keep the `dateFormat` in conditional rendering can stay because it is applied on a different field, some synchronisation is expected

**Review by weili-govtech — CHANGES_REQUESTED** [review]

File: src/components/fields/date-field/date-field.tsx (hunk 23 above)
Comment (hhung2710) [reply src/components/fields/date-field/date-field.tsx:57]: so to use different dateFormat dev need to defined synchronisation in both place right ?

File: src/components/fields/date-field/types.ts [hunk 24]
```diff
@@ -11,6 +12,7 @@ export interface IDateFieldValidationRule extends IYupValidationRule {
 	minDate?: string | undefined;
 	maxDate?: string | undefined;
 	excludedDates?: string[] | undefined;
+	withinDays?: TComponentOmitProps<IWithinDaysRule, "dateFormat"> | undefined;
```
Comment (weili-govtech) [inline src/components/fields/date-field/types.ts:15]: no need to use `TComponentOmitProps` because that is for omitting FEE schema keys

```suggestion
	withinDays?: Omit<IWithinDaysRule, "dateFormat"> | undefined;
```

**Review by weili-govtech — APPROVED** [review]
